src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/PostgresAccessPolicyDAO.scala

```diff
@@ -744,23 +744,53 @@ class PostgresAccessPolicyDAO(protected val writeDbRef: DbReference, protected v
   override def deleteAllResourcePolicies(resourceId: FullyQualifiedResourceId, samRequestContext: SamRequestContext): IO[Unit] = {
     val p = PolicyTable.syntax("p")
     val g = GroupTable.syntax("g")
+    val gm = GroupMemberTable.syntax("gm")
 
-    serializableWriteTransaction("deletePolicy", samRequestContext)({ implicit session =>
-      // first get all the group PKs associated to all the resource policies
-      // then delete all the resource policies
-      // then delete all the groups from the first query
-
-      val fromAndWhereFragment = samsqls"""from ${PolicyTable as p} where ${p.resourceId} = (${loadResourcePKSubQuery(resourceId)})"""
-
-      val groupPKsToDelete = samsql"""select ${p.result.groupId} $fromAndWhereFragment""".map(_.get[GroupPK](p.resultName.groupId)).list().apply()
-
-      samsql"""delete $fromAndWhereFragment""".update().apply()
+    val fromAndWhereFragment = samsqls"""from ${PolicyTable as p} where ${p.resourceId} = (${loadResourcePKSubQuery(resourceId)})"""
 
-      if (groupPKsToDelete.nonEmpty) {
-        samsql"""delete from ${GroupTable as g}
+    for {
+      groupPKsToDelete <- readOnlyTransaction("deletePolicyFindGroups", samRequestContext) { implicit session =>
+          samsql"""select ${p.result.groupId} $fromAndWhereFragment""".map(_.get[GroupPK](p.resultName.groupId)).list().apply()
+        }
+      writeResult <- serializableWriteTransaction("deletePolicy", samRequestContext)({ implicit session =>
+        // first get all the group PKs associated to all the resource policies
+        // then delete all the resource policies
+        // then delete all the groups from the first query
+        samsql"""delete $fromAndWhereFragment""".update().apply()
+
+        if (groupPKsToDelete.nonEmpty) {
+          samsql"""delete from ${GroupTable as g}
                  where ${g.id} in ($groupPKsToDelete)""".update().apply()
+        }
+      }).attempt
+      _ <- writeResult match {
+        case Left(fkViolation: PSQLException) if fkViolation.getSQLState == PSQLStateExtensions.FOREIGN_KEY_VIOLATION =>
+          for {
+            (problematicGroupIdsToNames, problematicGroupMemberships) <- readOnlyTransaction("deletePolicyErrorFindProblematicGroups", samRequestContext) { implicit session =>
+            val problematicGroupIdsToNames: Map[GroupPK, String] =
+              samsql"""select ${g.result.id}, ${g.result.name}
```

Comment:
These queries feel like they could be condensed into a single query -- possibly something like this? While we're grabbing all the problematic group ids and names, also grab all of the groups that they're a member of.

```
select ${g.result.id}, ${g.result.name}, ${pg.result.name}
                     from ${GroupTable as g}
                     join ${GroupMemberTable as gm} on ${g.id} = ${gm.memberGroupId}
                     join ${GroupTable as pg} on ${pg.id} = ${gm.groupId}
                     where ${g.id} in
                         (select distinct ${gm.result.memberGroupId}
                          from ${GroupMemberTable as gm}
                          where ${gm.memberGroupId} in ($groupPKsToDelete))
```

Reply:
+1

Reply:
+1